topsort-api-v2.yml

```diff
@@ -466,6 +508,206 @@ components:
+            This search string must match the searchQuery field that was provided in the auction request (if provided).
+          minLength: 1
+    EventsRequest:
+      type: object
+      properties:
+        impressions:
+          type: array
+          items:
+            $ref: '#/components/schemas/Impression'
+          minItems: 1
+          maxItems: 50
+        clicks:
+          type: array
+          items:
+            $ref: '#/components/schemas/Click'
+          minItems: 1
+          maxItems: 50
+        purchases:
+          type: array
+          items:
+            $ref: '#/components/schemas/Purchase'
+          minItems: 1
+          maxItems: 50
+      minProperties: 1
+      additionalProperties: false
+    Impression:
+      type: object
+      description: >
+        An impression means an ad has become visible to the consumer.
+        In case you cannot send an impression when the product becomes visible, send us an impression event when the product was rendered in the HTML or, if that's also not possible, when your API returns the results.
+        It is important to select the most specific event so that your vendors have more accurate CTR metrics, which will allow them to better predict their campaigns.
+      required:
+        - placement
+        - occurredAt
+        - opaqueUserId
+        - id
+      properties:
+        resolvedBidId:
+          type: string
+          description: If the impression is over an ad promotion, this is the `resolvedBidId` field received from the `/auctions` request.
+          example: WyJiX01mazE1IiwiMTJhNTU4MjgtOGVhZC00Mjk5LTgzMjctY2ViYjAwMmEwZmE4IiwibGlzdGluZ3MiLCJkZWZhdWx0IiwiIl0=
+        placement:
+          $ref: '#/components/schemas/Placement'
+        occurredAt:
+          type: string
+          format: date-time
+          description: RFC3339 formatted timestamp including UTC offset.
+          example: '2009-01-01T12:59:59-05:00'
+        opaqueUserId:
+          type: string
+          description: >
+            The opaque user ID allows correlating user activity, whether or not they are actually logged in.
+            It must be long lived.
+          example: 71303ce0-de89-496d-8270-6434589615e8
+        id:
+          type: string
+          description: >
+            The marketplace's unique ID for the impression.
+            This field ensures the event reporting is idempotent in case there is a network issue and the request is retried.
+            If there is no impression model on the marketplace side, generate a unique string that does not change if the event is resent.
+          example: eb874c98-bf4d-40a9-ae6d-fcf4cecb535c
+    Click:
+      description: >
+        A click is sent to Topsort when the consumer has clicked on an ad.
+        Topsort charges the vendor and pays the marketplace for clicks on ads in promoted placements on the marketplace app.
+      type: object
+      required:
+        - placement
```

Comment:
Why is the placement required?

Reply:
Should we leave it for impressions only?

Reply:
I meant in general, we are seeing now that marketplaces are sending invalid placement data, just to make the API happy.

We could not remove this data only from clicks unless they send us the corresponding impression id

Reply:
I see, so you're suggesting we let marketplaces skip the placement object and report it only if they will actually provide real information (and not a filler value to make the API happy).

Reply:
Yes. I guess though it's not that important, as nobody will be using this as this is only a preview.

Reply:
I've set it to optional. I think it's reasonable for marketplaces to assume this needs to be present if they want more advanced metrics over performance.